Custom Equipment Support

Jepvid opened the pull request:
Used https://github.com/HarbourMasters/Shipwright/pull/3932 as base
Gonna update hooks once https://github.com/HarbourMasters/Shipwright/pull/6026 is merged

_Commit history is wierd as i used github actions for building and debugging_

## Files

https://github.com/HarbourMasters/Shipwright/files/14305452/Alt_Equip_Template.zip

Download template mod file from this link and mess around

If you wish to make your own mod to test this

https://docs.google.com/spreadsheets/d/1rOTt_7Wr0OGfMR9tHom8dDOooM3rUocz9xi7axpR0sY/edit#gid=556482990

Export to the items to `objects/object_custom_equip` following Dana's spreadsheet

## What this pr does

This PR implements object_custom_equip objects that adds more support to have items modified by modders without having to export with hands or do xml edits to load hands. This PR will make this autonomous if exported as a object_custom_equip.

This pr will also add non existing DisplayList for items that originally did not have a DisplayList. Longshot and Big Goron Sword Sheath is 2 examples

## Summary of what I did

Changes make patching safe and idempotent  when toggling Alt Assets and when mods overwrite assets.

Moved hooks to use ShipInit to work on the assets not loading issue.

New function made with help of chatgpt in order to protect vanilla assets from ever being unloaded.
Creates a runtime alt DisplayList by cloning the vanilla asset if no alt asset exist, enabling safe custom equipment patching.
[Link to discord message explaining the reason why this function exist](https://discord.com/channels/808039310850130000/1161885132911816724/1451336437646753914 )

ChatGPT was used during troubleshooting.

## Important Info

From Pats PR

> Known quirk is that both a Custom Shield and Sword/Sheath must be present or it will revert to the vanilla DL, because I can't just patch one and not everything will be loaded properly. i.e. If you have just a custom Master Sword, but no custom Hylian Shield, it would not be able to render a shield at all, because I can't patch into vanilla shield DL's, especially if another mod is being used. So one of each must be present.


## Known Issues

- ~Models ocassionally reverts back to normal assets~ Fixed with new hooks and ShipInit
- ~sometimes when alt toggling vanilla assets unload completely~ Fixed by implementing a function to protect vanilla assets from unload shenanigans
- ~when child Link is using adult items with `Scale Adult Equiment As Child` enabled. Hands get scaled down~ Not an issue from this pr. Is just part of what that cvar does, in order to make it proper with custom equipment a refactor of the cvar may be in order.
- ~When child and swordless, if Link has a shield it will display master sword on the back with dekushield, working on a fix~ Fix implemented, tried my best for timeless support but there are some edgecases that seems more engine issue than custom equipment. Vanilla combo is working as expected

## Small word of thanks

Thanks to @aMannus and @Patrick12115 For giving me helping hand either directly or indirectly


<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4939591612.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4939605104.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4939669791.zip)
<!--- section:artifacts:end -->

Jepvid committed:
Add custom equipment display lists header

This header file defines custom equipment display lists for various items, including weapons and shields, with appropriate alignment macros.

Jepvid committed:
Add ResourceMgr_PatchCustomGfxByName function

Jepvid committed:
Add ResourceMgr_PatchCustomGfxByName function

Jepvid committed:
Add custom equipment patching functionality

This file implements custom equipment patching for various items, including swords, shields, and ocarinas, based on player context and equipped items. It registers hooks to update the display lists for custom equipment dynamically.

Jepvid committed:
Implement custom scaling for Deku Shield on child

Added support for custom scaling of the Deku Shield when the player is a child.

Jepvid committed:
Include GameInteractor and ResourceManager headers

Added includes for GameInteractor and ResourceManager.

Jepvid committed:
Fix include path for ResourceManager header

Jepvid committed:
Replace ResourceManager include with ResourceManagerHelpers

Jepvid committed:
Fix include path for ResourceManagerHelpers header

Jepvid committed:
Implement equipment visibility updates for child and adult

Jepvid committed:
Add custom equipment Dlists handler registration

Jepvid committed:
Add custom equipment header to mods.cpp

Jepvid committed:
Add custom equipment header to mods.cpp

Jepvid committed:
Add include for custom equipment header

Jepvid committed:
Add custom equipment header with registration function

Jepvid committed:
Add include guards to customequipment.h

Jepvid committed:
Add function declaration for RegisterPatchCustomEquipmentDlistsHandler

Jepvid committed:
Add UpdatePatchCustomEquipmentDlists function declaration

Jepvid committed:
Remove UpdatePatchCustomEquipmentDlists declaration

Removed the declaration of UpdatePatchCustomEquipmentDlists function.

Jepvid committed:
Remove custom equipment include from mods.cpp

Removed unnecessary include for custom equipment.

Jepvid committed:
Remove custom equipment include from mods.cpp

Removed unnecessary include for custom equipment.

Jepvid committed:
Add ShipInit.hpp include to customequipment.cpp

Jepvid committed:
Add custom equipment header to mods.cpp

Jepvid committed:
Add Patch Hand Handler registration

Jepvid committed:
Add RegisterPatchHandHandler function declaration

Jepvid committed:
Call UpdatePatchCustomEquipmentDlists after registration

Jepvid committed:
Change enhancement setting check in PatchOrUnpatch

Jepvid committed:
Change setting check from 'EnabledMods' to 'AltAssets'

Jepvid committed:
Call UpdatePatchCustomEquipmentDlists after registration

Jepvid committed:
Fix registration of OnAssetAltChange hook

Jepvid committed:
Update customequipment.cpp

Jepvid committed:
Update customequipment.cpp

Jepvid committed:
Update customequipment.cpp

Jepvid committed:
Update customequipment.cpp

Jepvid committed:
Update ResourceManagerHelpers.cpp

Jepvid committed:
Update Custom Equipment (#1)

* Update customequipment.cpp

* Shipinit and 2 new hooks

* Fix Build error

* Cleanup ShipInit

* Come on Shipinit, work...

* Woops

* Error between user and keyboard

* Forgot this existed

* Add Enter FPS hook

* Remove "duplicate" hook and test disable extra hook

* Vanilla Fix please, i want vanilla

* I said i wanted vanilla

Jepvid committed:
Update Custom Equipment (#1)

* Update customequipment.cpp

* Shipinit and 2 new hooks

* Fix Build error

* Cleanup ShipInit

* Come on Shipinit, work...

* Woops

* Error between user and keyboard

* Forgot this existed

* Add Enter FPS hook

* Remove "duplicate" hook and test disable extra hook

* Vanilla Fix please, i want vanilla

* I said i wanted vanilla

Jepvid committed:
Alt equipment Unstable Build (#3)

* Update customequipment.cpp

* Shipinit and 2 new hooks

* Fix Build error

* Cleanup ShipInit

* Come on Shipinit, work...

* Woops

* Error between user and keyboard

* Forgot this existed

* Add Enter FPS hook

* Remove "duplicate" hook and test disable extra hook

* Vanilla Fix please, i want vanilla

* I said i wanted vanilla

* Scaling Cleanup

* Remove cond_hook

* Lets Pray

* FINAL PRAYER!

* go back!

* GO BACK MORE

Jepvid committed:
Alt equipment Cleanup (#4)

* Update ResourceManagerHelpers.cpp

* Electric Boogalooooo

* Revert Resourcemanager and make "revised" resourcemanager function for custom

* Forgot header file....

* Revert more stuff

* Hookshot better be hooking now

* hoooooks

* wooops

* scatterbrain moment

* Forceloadhookshot dl

* i forgor

* Revert last stuff and cleanup

* Dont give custom equipment special treatment

* git failed

* lets try this style

* new try

* Helpers Hope

* I believe in this one

* No helpers this time

* once more

* Im at the tip of this hookshot

* im tired

* Im feeling fancy

* Cleanslate

Jepvid committed:
Alt equipment (#5)

* Another layer

* Fux sake

* really?

* again?

Jepvid committed:
Clang time

garrettjoecox commented:
Would you be able to provide an o2r that includes all of these items? Doesn't need to be pretty just would be good to have for validating, and I'd like to experiment with some different patching patterns if I find some time.

Jepvid commented:
> Would you be able to provide an o2r that includes all of these items? Doesn't need to be pretty just would be good to have for validating, and I'd like to experiment with some different patching patterns if I find some time.

It is packed in the Alt_Equipment_Template zip file in the description. Its provided by Dana

garrettjoecox commented:
> It is packed in the Alt_Equipment_Template zip file in the description. Its provided by Dana

Wow my eyes must've glazed right over that one, thanks

garrettjoecox commented on soh/src/code/z_player_lib.c:
```diff
@@ -5,6 +5,7 @@
 #include "objects/object_link_child/object_link_child.h"
 #include "objects/object_triforce_spot/object_triforce_spot.h"
 #include "overlays/actors/ovl_Demo_Effect/z_demo_effect.h"
+#include "objects/object_custom_equip/object_custom_equip.h"
```
```suggestion
```

garrettjoecox commented on soh/src/code/z_player_lib.c:
```diff
@@ -1481,7 +1483,7 @@ s32 Player_OverrideLimbDrawGameplayFirstPerson(PlayState* play, s32 limbIndex, G
         }
     }
-
+    GameInteractor_ExecuteOnPlayerAiming();
```
Doesn't look like this is actually used by anything you've added
```suggestion
```

Jepvid replied:
Remnants from when fps item were not loading properly. Didnt use as it was ran every frame

garrettjoecox commented on soh/src/code/z_player_lib.c:
```diff
@@ -650,6 +651,7 @@ void Player_SetModels(Player* this, s32 modelGroup) {
     this->waistDLists = &sPlayerDListGroups[gPlayerModelTypes[modelGroup][4]][gSaveContext.linkAge];
 
     Player_SetModelsForHoldingShield(this);
+    GameInteractor_ExecuteOnPlayerChangeItem();
```
Not necessarily only/always called when changing items, this would be a more appropriate name
```suggestion
    GameInteractor_ExecuteOnPlayerSetModels(modelGroup);
```

garrettjoecox commented on soh/soh/ResourceManagerHelpers.cpp:
```diff
@@ -357,6 +357,71 @@ extern "C" void ResourceMgr_PatchGfxByName(const char* path, const char* patchNa
 }
 
+// Create or get existing alt DisplayList for custom equipment
+static std::shared_ptr<Fast::DisplayList> ResourceMgr_GetOrCreateAltDisplayList(const char* path);
```
```suggestion
```

garrettjoecox commented on soh/soh/Enhancements/customequipment.h:
```diff
@@ -0,0 +1,7 @@
+void RegisterPatchCustomEquipmentDlistsHandler();
+void RegisterPatchHandHandler();
+
+#endif // CUSTOM_EQUIP_H
```
Can probably just delete this whole file
```suggestion
```

garrettjoecox commented on soh/soh/Enhancements/game-interactor/GameInteractor_Hooks.cpp:
```diff
@@ -189,10 +189,18 @@ void GameInteractor_ExecuteOnTimestamp(u8 item) {
     GameInteractor::Instance->ExecuteHooks<GameInteractor::OnTimestamp>(item);
 }
 
+void GameInteractor_ExecuteOnPlayerAiming() {
+    GameInteractor::Instance->ExecuteHooks<GameInteractor::OnPlayerAiming>();
+}
+
```
```suggestion
```

garrettjoecox commented on soh/soh/Enhancements/game-interactor/GameInteractor_Hooks.h:
```diff
@@ -41,7 +41,9 @@ void GameInteractor_ExecuteOnActorDestroy(void* actor);
 void GameInteractor_ExecuteOnEnemyDefeat(void* actor);
 void GameInteractor_ExecuteOnBossDefeat(void* actor);
 void GameInteractor_ExecuteOnTimestamp(u8 item);
+void GameInteractor_ExecuteOnPlayerAiming();
```
```suggestion
```

garrettjoecox commented on soh/soh/Enhancements/game-interactor/GameInteractor_HookTable.h:
```diff
@@ -40,7 +40,9 @@ DEFINE_HOOK(OnActorDestroy, (void* actor));
 DEFINE_HOOK(OnEnemyDefeat, (void* actor));
 DEFINE_HOOK(OnBossDefeat, (void* actor));
 DEFINE_HOOK(OnTimestamp, (u8 item));
+DEFINE_HOOK(OnPlayerAiming, ());
```
```suggestion
```

Jepvid committed:
BugFix for swordless child link (some timeless support) (#6)

* BugFix for child link back sword

* Second try

* Test this

* Another one

* Getting close

* Timeless support

* Sanity check

* Final patch for timeless

Jepvid commented:
Added fix for some swordless issues that passed through initial testing.

Jepvid committed:
Alt equipment suggestions (#7)

* first pass

* im dumb

* i forgor

* seriously?

* Hopefully lastone

* frogot

* clang

Jepvid commented:
Implemented suggestions

Jepvid committed:
Optimize and make customequip.cpp more readable

Applied fix to a crash when changing age

Jepvid committed:
clang

Jepvid committed:
Final Fixes

Jepvid committed:
forgot to save

Jepvid committed:
clang clang

serprex committed:
Refactor Forest (#5673)

* refactor forest

* feedback

* more feedback

* MQ NE island logic

* no kids allowed

* split west corridor

* split up block puzzle room

* comment mapping poe sister colors

* generalize well swim, make novice

* hammer works

* upstream bug fix

* Replace corridor with hallway, rename east/west hallway regions to be based on their red/blue doormats

also change south hallway to overgrown hallway, & replace south/north sides with lower/upper sides

Jepvid committed:
one line.....

Jepvid committed:
Optimize and fix crash

Optimize and make customequip.cpp more readable

Fix crash that could happen with age changing when using custom equipment

Jepvid commented:
Final Cleanup done. I aint touching this anymore

Jepvid committed:
Thx pat!

Jepvid committed:
Thx Pat! Hotfix

Jepvid committed:
I shouldnt have done that

Removed a line along time ago that was being used for something that wasnt apparent on my end

Jepvid commented:
Marking this as draft for now as Ive gotten reports of its not working for others. Debugging time

Jepvid committed:
Update CVAR to correct CVAR

* FOUND IT

* Cleanup

Jepvid commented:
Found what was causing issues. Fix have been applied

PR is ready for review

garrettjoecox committed:
Fix rupee overflow in StartingItemGive (#6039)

jeromkiller committed:
Reverted removal of 'ClearItemLocations' to fix #6036 (#6044)

serprex committed:
Fix mixed up chest names in shadow MQ falling spikes room (#6047)

serprex committed:
refactor windmill (#6027)

split upper area into region rather than using logic val to manage access from below

2 fixes: adult can groundjump up the middle, child can make the trick with hammer

serprex committed:
Fairysanity: allow using bottle (#6021)

Based on fishsanity, which is refactored with hooks

Also open up logic catching fairies at oasis if player has bottle

serprex committed:
RT_BIG_SKULLTULA_PAUSE_LIFT (#6014)

serprex committed:
RT_WATER_IRON_BOOTS_LEDGE_GRAB (#6009)

https://www.youtube.com/watch?v=x8k3XJHiMOs

Jepvid committed:
Fixed Crashes

* Otherversion

* Final Build for reals

* Cleanup

* Debuggers

* clang

* Update customequipment.cpp

* Fix Fallback for BGS

* Fix anchor crash

* Clang

Jepvid commented:
Fixed the final crashes made with anchor now. This pr should be ready for review now.

nclok1405 committed:
Add Gerudo Fighter to Enemy Randomizer (#6005)

Being captured by random fighter behaves like being captured by a wall master

serprex committed:
RT_ICE_STALAGMITE_CLIP, RT_ICE_STALAGMITE_HOOKSHOT (#5909)

Coming back from block room to hub, the clip is different, instead it's by wall that both ages can get past with pretty specific sidehop. This is pretty niche since generally logic isn't doing reverse ice cavern

serprex committed:
bushsanity (#5941)

serprex committed:
more perm flags (#5676)

1. deku tree MQ web
2. dodongo's cavern MQ silver rupees
3. jabu lobby switch
4. forest well, lobby eyes, MQ web
5. fire MQ lobby torches
6. spirit MQ lobby silver rupees

adjusts forest MQ logic

xxAtrain223 committed:
Add GetCheckPrice (#5718)

* Removed LocationAccess::CanBuy and added GetCheckPrice.

* Made the Magic Bean Salesman a merchant.

* Replaced CanBuyAnother with GetWalletCapacity and inline comparison for easier readability.

aMannus committed:
Fix custom tunics not updating on scene change (#6026)

JordanLongstaff committed:
Convert "Big Four" hook register functions to init functions (#5984)

OtherBlue committed:
[Enhancement] Unequip C-items (#6043)

serprex committed:
Shuffle Bean Souls (#5833)

serprex committed:
RT_GTG_LAVA_JUMP (#5987)

Distinct from unintuitive jumps in that I'm unable to get it consistently even after trying a couple times,
even after trying to find some mix of ledge grabbing & ESS turning to try at least put it down to roll timing

serprex committed:
Add missing extension trick to Shadow MQ, fix 2 logic mixups in GTG (#5983)

Also allow Giant's Knife to logically hit these switches

Ganon's Castle diamond switch can be hit with bomb by setting bomb down instead of shield dropping

Jepvid committed:
Update to latest and Updatehooks

* updatehooks

* fix crash

Jepvid committed:
Update to latest and Updatehooks

* updatehooks

* fix crash

Jepvid committed:
Revert "Update to latest and Updatehooks"

This reverts commit cb0f8802b16c95150a168b1890d7ac8a4f86c793.

Jepvid committed:
Revert "Update to latest and Updatehooks"

This reverts commit cb0f8802b16c95150a168b1890d7ac8a4f86c793.

Jepvid commented:
Closing this as it got nuked to hell trying to rebase this
